File: src/main/java/org/profit/server/handler/HttpStaticFileHandler.java
package org.profit.server.handler;

import io.netty.buffer.Unpooled;
import io.netty.channel.*;
import io.netty.handler.codec.http.*;
import io.netty.util.CharsetUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.activation.MimetypesFileTypeMap;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.RandomAccessFile;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.regex.Pattern;

import static io.netty.handler.codec.http.HttpHeaders.Names.*;
import static io.netty.handler.codec.http.HttpHeaders.isKeepAlive;
import static io.netty.handler.codec.http.HttpHeaders.setContentLength;
import static io.netty.handler.codec.http.HttpMethod.GET;
import static io.netty.handler.codec.http.HttpResponseStatus.*;
import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;

public class HttpStaticFileHandler {

	private static final Logger LOG = LoggerFactory.getLogger(HttpStaticFileHandler.class);

	private static final MimetypesFileTypeMap MIMETYPES_FILE_TYPE_MAP;

	public static final String HTTP_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss zzz";
	public static final String HTTP_DATE_GMT_TIMEZONE = "GMT";
	public static final int HTTP_CACHE_SECONDS = 60;
	public static final String WEB_ROOT = "webroot";

	private static final Pattern INSECURE_URI = Pattern.compile(".*[<>&\"].*");

	static {
		MIMETYPES_FILE_TYPE_MAP = new MimetypesFileTypeMap();
		MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/xml xml");
		MIMETYPES_FILE_TYPE_MAP.addMimeTypes("text/html html");
		MIMETYPES_FILE_TYPE_MAP.addMimeTypes("text/css css");
		MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/javascript js");
		MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/json json");
	}

	public void handleStaticFile(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
		if (!request.getDecoderResult().isSuccess()) {
			sendError(ctx, BAD_REQUEST);
			return;
		}

		if (request.getMethod() != GET) {
			sendError(ctx, METHOD_NOT_ALLOWED);
			return;
		}

		QueryStringDecoder decoder = new QueryStringDecoder(request.getUri());
		String uri = decoder.path();
		if (uri.equals("/")) {
			uri = "/index.html";
		} else if (uri.endsWith("en")) {
			uri = "/index-en.html";
		}

//        QueryStringDecoder decoder = new QueryStringDecoder(request.getUri());
//        final String uri = decoder.path();
//		final String path = sanitizeUri(uri);
//		if (path == null) {
//			sendError(ctx, FORBIDDEN);
//			return;
//		}

		final String path = sanitizeUri(uri);
		if (path == null) {
			sendError(ctx, FORBIDDEN);
			return;
		}

		LOG.debug("Request file path: {}", path);

		File file = new File(path);
		if (file.isHidden() || !file.exists()) {
			sendError(ctx, NOT_FOUND);
			return;
		}
		if (!file.isFile()) {
			sendError(ctx, FORBIDDEN);
			return;
		}

		// Cache Validation
		String ifModifiedSince = request.headers().get(IF_MODIFIED_SINCE);
		if (ifModifiedSince != null && !ifModifiedSince.isEmpty()) {
			SimpleDateFormat dateFormatter = new SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US);
			Date ifModifiedSinceDate = dateFormatter.parse(ifModifiedSince);

			// Only compare up to the second because the datetime format we send to the client
			// does not have milliseconds
			long ifModifiedSinceDateSeconds = ifModifiedSinceDate.getTime() / 1000;
			long fileLastModifiedSeconds = file.lastModified() / 1000;
			if (ifModifiedSinceDateSeconds == fileLastModifiedSeconds) {
				sendNotModified(ctx);
				return;
			}
		}

		RandomAccessFile raf;
		try {
			raf = new RandomAccessFile(file, "r");
		} catch (FileNotFoundException fnfe) {
			sendError(ctx, NOT_FOUND);
			return;
		}

		long fileLength = raf.length();
		HttpResponse response = new DefaultHttpResponse(HTTP_1_1, OK);
		setContentLength(response, fileLength);
		setContentTypeHeader(response, file);
		setDateAndCacheHeaders(response, file);
		if (isKeepAlive(request)) {
			response.headers().set(CONNECTION, HttpHeaders.Values.KEEP_ALIVE);
		}

		// Write the initial line and the header.
		ctx.write(response);

		ChannelFuture sendFileFuture = ctx.write(new DefaultFileRegion(raf.getChannel(), 0, fileLength), ctx.newProgressivePromise());
		sendFileFuture.addListener(new ChannelProgressiveFutureListener() {
			@Override
			public void operationProgressed(ChannelProgressiveFuture future, long progress, long total) {
				if (total < 0) { // total unknown
					LOG.debug("Transfer progress: {}", progress);
				} else {
					LOG.debug("Transfer progress: {}", progress + " / " + total);
				}
			}

			@Override
			public void operationComplete(ChannelProgressiveFuture future) throws Exception {
				LOG.debug("Transfer complete.");
			}
		});

		// Write the end marker
		ChannelFuture lastContentFuture = ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);

		// Decide whether to close the connection or not.
		if (!isKeepAlive(request)) {
			// Close the connection when the whole content is written out.
			lastContentFuture.addListener(ChannelFutureListener.CLOSE);
		}
	}

	private static String sanitizeUri(String uri) {
		// Decode the path.
		try {
			uri = URLDecoder.decode(uri, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			try {
				uri = URLDecoder.decode(uri, "ISO-8859-1");
			} catch (UnsupportedEncodingException e1) {
				throw new Error();
			}
		}

		if (!uri.startsWith("/")) {
			return null;
		}

		// Convert file separators.
		uri = uri.replace('/', File.separatorChar);

		// Simplistic dumb security check.
		// You will have to do something serious in the production environment.
		if (uri.contains(File.separator + '.') ||
				uri.contains('.' + File.separator) ||
				uri.startsWith(".") || uri.endsWith(".") ||
				INSECURE_URI.matcher(uri).matches()) {
			return null;
		}

		// Convert to absolute path.
        String path = System.getProperty("user.dir") + File.separator + WEB_ROOT + File.separator + uri;
		return path;
	}

	private static void sendRedirect(ChannelHandlerContext ctx, String newUri) {
		FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, FOUND);
		response.headers().set(LOCATION, newUri);

		// Close the connection as soon as the error message is sent.
		ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
	}

	private static void sendError(ChannelHandlerContext ctx, HttpResponseStatus status) {
		FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, status, Unpooled.copiedBuffer("Failure: " + status.toString() + "\r\n", CharsetUtil.UTF_8));
		response.headers().set(CONTENT_TYPE, "text/plain; charset=UTF-8");

		// Close the connection as soon as the error message is sent.
		ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
	}

	/**
	 * When file timestamp is the same as what the browser is sending up, send a "304 Not Modified"
	 *
	 * @param ctx Context
	 */
	private static void sendNotModified(ChannelHandlerContext ctx) {
		FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, NOT_MODIFIED);
		setDateHeader(response);

		// Close the connection as soon as the error message is sent.
		ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
	}

	/**
	 * Sets the Date header for the HTTP response
	 *
	 * @param response HTTP response
	 */
	private static void setDateHeader(FullHttpResponse response) {
		SimpleDateFormat dateFormatter = new SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US);
		dateFormatter.setTimeZone(TimeZone.getTimeZone(HTTP_DATE_GMT_TIMEZONE));

		Calendar time = new GregorianCalendar();
		response.headers().set(DATE, dateFormatter.format(time.getTime()));
	}

	/**
	 * Sets the Date and Cache headers for the HTTP Response
	 *
	 * @param response    HTTP response
	 * @param fileToCache file to extract content type
	 */
	private static void setDateAndCacheHeaders(HttpResponse response, File fileToCache) {
		SimpleDateFormat dateFormatter = new SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US);
		dateFormatter.setTimeZone(TimeZone.getTimeZone(HTTP_DATE_GMT_TIMEZONE));

		// Date header
		Calendar time = new GregorianCalendar();
		response.headers().set(DATE, dateFormatter.format(time.getTime()));

		// Add cache headers
		time.add(Calendar.SECOND, HTTP_CACHE_SECONDS);
		response.headers().set(EXPIRES, dateFormatter.format(time.getTime()));
		response.headers().set(CACHE_CONTROL, "private, max-age=" + HTTP_CACHE_SECONDS);
		response.headers().set(LAST_MODIFIED, dateFormatter.format(new Date(fileToCache.lastModified())));
	}

	/**
	 * Sets the content type header for the HTTP Response
	 *
	 * @param response HTTP response
	 * @param file     file to extract content type
	 */
	private static void setContentTypeHeader(HttpResponse response, File file) {
		response.headers().set(CONTENT_TYPE, MIMETYPES_FILE_TYPE_MAP.getContentType(file.getPath()));
	}
}

File: src/main/java/org/profit/server/handler/HttpStaticFileHandler.kt
package org.profit.server.handler

import io.netty.buffer.Unpooled
import io.netty.channel.*
import io.netty.handler.codec.http.*
import io.netty.util.CharsetUtil
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileNotFoundException
import java.io.RandomAccessFile
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern
import javax.activation.MimetypesFileTypeMap

class HttpStaticFileHandler {
    companion object {
        private val LOG = LoggerFactory.getLogger(HttpStaticFileHandler::class.java)

        val MIMETYPES_FILE_TYPE_MAP: MimetypesFileTypeMap = MimetypesFileTypeMap()

        const val HTTP_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss zzz"
        const val HTTP_DATE_GMT_TIMEZONE = "GMT"
        const val HTTP_CACHE_SECONDS = 60
        const val WEB_ROOT = "webroot"
        private val INSECURE_URI = Pattern.compile(".*[<>&\"].*")
        private fun sanitizeUri(uri: String): String? { // Decode the path.
            var uri = uri
            uri = try {
                URLDecoder.decode(uri, "UTF-8")
            } catch (e: UnsupportedEncodingException) {
                try {
                    URLDecoder.decode(uri, "ISO-8859-1")
                } catch (e1: UnsupportedEncodingException) {
                    throw Error()
                }
            }
            if (!uri.startsWith("/")) {
                return null
            }
            // Convert file separators.
            uri = uri.replace('/', File.separatorChar)
            // Simplistic dumb security check.
// You will have to do something serious in the production environment.
            return if (uri.contains(File.separator + '.') ||
                    uri.contains('.'.toString() + File.separator) ||
                    uri.startsWith(".") || uri.endsWith(".") ||
                    INSECURE_URI.matcher(uri).matches()) {
                null
            } else System.getProperty("user.dir") + File.separator + WEB_ROOT + File.separator + uri
            // Convert to absolute path.
        }

        private fun sendRedirect(ctx: ChannelHandlerContext, newUri: String) {
            val response: FullHttpResponse = DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.FOUND)
            response.headers()[HttpHeaders.Names.LOCATION] = newUri
            // Close the connection as soon as the error message is sent.
            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
        }

        private fun sendError(ctx: ChannelHandlerContext, status: HttpResponseStatus) {
            val response: FullHttpResponse = DefaultFullHttpResponse(HttpVersion.HTTP_1_1, status, Unpooled.copiedBuffer("Failure: $status\r\n", CharsetUtil.UTF_8))
            response.headers()[HttpHeaders.Names.CONTENT_TYPE] = "text/plain; charset=UTF-8"
            // Close the connection as soon as the error message is sent.
            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
        }

        /**
         * When file timestamp is the same as what the browser is sending up, send a "304 Not Modified"
         *
         * @param ctx Context
         */
        private fun sendNotModified(ctx: ChannelHandlerContext) {
            val response: FullHttpResponse = DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_MODIFIED)
            setDateHeader(response)
            // Close the connection as soon as the error message is sent.
            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
        }

        /**
         * Sets the Date header for the HTTP response
         *
         * @param response HTTP response
         */
        private fun setDateHeader(response: FullHttpResponse) {
            val dateFormatter = SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US)
            dateFormatter.timeZone = TimeZone.getTimeZone(HTTP_DATE_GMT_TIMEZONE)
            val time: Calendar = GregorianCalendar()
            response.headers()[HttpHeaders.Names.DATE] = dateFormatter.format(time.time)
        }

        /**
         * Sets the Date and Cache headers for the HTTP Response
         *
         * @param response    HTTP response
         * @param fileToCache file to extract content type
         */
        private fun setDateAndCacheHeaders(response: HttpResponse, fileToCache: File) {
            val dateFormatter = SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US)
            dateFormatter.timeZone = TimeZone.getTimeZone(HTTP_DATE_GMT_TIMEZONE)
            // Date header
            val time: Calendar = GregorianCalendar()
            response.headers()[HttpHeaders.Names.DATE] = dateFormatter.format(time.time)
            // Add cache headers
            time.add(Calendar.SECOND, HTTP_CACHE_SECONDS)
            response.headers()[HttpHeaders.Names.EXPIRES] = dateFormatter.format(time.time)
            response.headers()[HttpHeaders.Names.CACHE_CONTROL] = "private, max-age=$HTTP_CACHE_SECONDS"
            response.headers()[HttpHeaders.Names.LAST_MODIFIED] = dateFormatter.format(Date(fileToCache.lastModified()))
        }

        /**
         * Sets the content type header for the HTTP Response
         *
         * @param response HTTP response
         * @param file     file to extract content type
         */
        private fun setContentTypeHeader(response: HttpResponse, file: File) {
            response.headers()[HttpHeaders.Names.CONTENT_TYPE] = MIMETYPES_FILE_TYPE_MAP!!.getContentType(file.path)
        }

        init {
            MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/xml xml")
            MIMETYPES_FILE_TYPE_MAP.addMimeTypes("text/html html")
            MIMETYPES_FILE_TYPE_MAP.addMimeTypes("text/css css")
            MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/javascript js")
            MIMETYPES_FILE_TYPE_MAP.addMimeTypes("application/json json")
        }
    }

    @Throws(Exception::class)
    fun handleStaticFile(ctx: ChannelHandlerContext, request: FullHttpRequest) {
        if (!request.decoderResult.isSuccess) {
            sendError(ctx, HttpResponseStatus.BAD_REQUEST)
            return
        }
        if (request.method !== HttpMethod.GET) {
            sendError(ctx, HttpResponseStatus.METHOD_NOT_ALLOWED)
            return
        }
        val decoder = QueryStringDecoder(request.uri)
        var uri = decoder.path()
        if (uri == "/") {
            uri = "/index.html"
        } else if (uri.endsWith("en")) {
            uri = "/index-en.html"
        }
        //        QueryStringDecoder decoder = new QueryStringDecoder(request.getUri());
//        final String uri = decoder.path();
//		final String path = sanitizeUri(uri);
//		if (path == null) {
//			sendError(ctx, FORBIDDEN);
//			return;
//		}
        val path = sanitizeUri(uri)
        if (path == null) {
            sendError(ctx, HttpResponseStatus.FORBIDDEN)
            return
        }
        LOG.debug("Request file path: {}", path)
        val file = File(path)
        if (file.isHidden || !file.exists()) {
            sendError(ctx, HttpResponseStatus.NOT_FOUND)
            return
        }
        if (!file.isFile) {
            sendError(ctx, HttpResponseStatus.FORBIDDEN)
            return
        }
        // Cache Validation
        val ifModifiedSince = request.headers()[HttpHeaders.Names.IF_MODIFIED_SINCE]
        if (ifModifiedSince != null && !ifModifiedSince.isEmpty()) {
            val dateFormatter = SimpleDateFormat(HTTP_DATE_FORMAT, Locale.US)
            val ifModifiedSinceDate = dateFormatter.parse(ifModifiedSince)
            // Only compare up to the second because the datetime format we send to the client
// does not have milliseconds
            val ifModifiedSinceDateSeconds = ifModifiedSinceDate.time / 1000
            val fileLastModifiedSeconds = file.lastModified() / 1000
            if (ifModifiedSinceDateSeconds == fileLastModifiedSeconds) {
                sendNotModified(ctx)
                return
            }
        }
        val raf: RandomAccessFile
        raf = try {
            RandomAccessFile(file, "r")
        } catch (fnfe: FileNotFoundException) {
            sendError(ctx, HttpResponseStatus.NOT_FOUND)
            return
        }
        val fileLength = raf.length()
        val response: HttpResponse = DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK)
        HttpHeaders.setContentLength(response, fileLength)
        setContentTypeHeader(response, file)
        setDateAndCacheHeaders(response, file)
        if (HttpHeaders.isKeepAlive(request)) {
            response.headers()[HttpHeaders.Names.CONNECTION] = HttpHeaders.Values.KEEP_ALIVE
        }
        // Write the initial line and the header.
        ctx.write(response)
        val sendFileFuture = ctx.write(DefaultFileRegion(raf.channel, 0, fileLength), ctx.newProgressivePromise())
        sendFileFuture.addListener(object : ChannelProgressiveFutureListener {
            override fun operationProgressed(future: ChannelProgressiveFuture, progress: Long, total: Long) {
                if (total < 0) { // total unknown
                    LOG.debug("Transfer progress: {}", progress)
                } else {
                    LOG.debug("Transfer progress: {}", "$progress / $total")
                }
            }

            @Throws(Exception::class)
            override fun operationComplete(future: ChannelProgressiveFuture) {
                LOG.debug("Transfer complete.")
            }
        })
        // Write the end marker
        val lastContentFuture = ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT)
        // Decide whether to close the connection or not.
        if (!HttpHeaders.isKeepAlive(request)) { // Close the connection when the whole content is written out.
            lastContentFuture.addListener(ChannelFutureListener.CLOSE)
        }
    }
}